core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala

```diff
@@ -416,6 +423,35 @@ class ExecutorSuite extends SparkFunSuite
     assert(taskMetrics.getMetricValue("JVMHeapMemory") > 0)
   }
 
+  test("SPARK-34949: do not re-register BlockManager when executor is shutting down") {
+    val reregisterInvoked = new AtomicBoolean(false)
+    val mockBlockManager = mock[BlockManager]
+    when(mockBlockManager.reregister()).thenAnswer { (_: InvocationOnMock) =>
+      reregisterInvoked.getAndSet(true)
+    }
+    val conf = new SparkConf(false).setAppName("test").setMaster("local[2]")
+    val mockEnv = createMockEnv(conf, new JavaSerializer(conf))
+    when(mockEnv.blockManager).thenReturn(mockBlockManager)
+
+    withExecutor("id", "localhost", mockEnv) { executor =>
+      withMockHeartbeatReceiverRef(executor) { mockReceiverRef =>
+        when(mockReceiverRef.askSync(any[Heartbeat], any[RpcTimeout])(any)).thenAnswer {
+          (_: InvocationOnMock) => HeartbeatResponse(reregisterBlockManager = true)
+        }
+        val reportHeartbeat = PrivateMethod[Unit](Symbol("reportHeartBeat"))
+        executor.invokePrivate(reportHeartbeat())
+        assert(reregisterInvoked.get(),
+          "BlockManager.reregister not invoked when reregisterBlockManager was true")
```

Comment:
@sumeetgajjar . Is this correct, `BlockManager.reregister not invoked`?

Reply:
Hi @dongjoon-hyun ,

I am slightly confused here about the question.
if you are pointing to the assertion then yes, I am asserting that when the executor shutdown is not initiated, `BlockManager.reregister` is invoked successfully by `Executor.reportHeartbeat`. 

However, if you pointing to the `assertion clue`, then I can see how that can be ambiguous, I can change the message to 
"BlockManager.reregister not invoked on Heartbeat response when executor is not shutting down"

Reply:
I am assuming @dongjoon-hyun meant to change the message around - "expected foo when bar but was baz".
As stated currently, it is a bit confusing, and I had to double check the logic as well

Reply:
Yes, I meant the message (assertion clue).

Reply:
Thank you for the clarification guys, have changed the message and updated the PR 😄.